hazelcast/src/main/java/com/hazelcast/internal/affinity/ThreadAffinitySupport.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.affinity;
+
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.spi.properties.GroupProperty;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.hazelcast.internal.affinity.ThreadAffinity.Group.IO;
+import static com.hazelcast.internal.affinity.ThreadAffinity.Group.PARTITION_THREAD;
+import static com.hazelcast.internal.affinity.ThreadAffinityProperties.isAffinityEnabled;
+import static com.hazelcast.util.ConcurrencyUtil.getOrPutIfAbsent;
+import static com.hazelcast.util.EmptyStatement.ignore;
+import static com.hazelcast.util.OsHelper.OS;
+import static com.hazelcast.util.OsHelper.isUnixFamily;
+import static com.hazelcast.util.Preconditions.checkNotNull;
+import static java.util.Collections.disjoint;
+
+public class ThreadAffinitySupport {
+
+    public static volatile ThreadAffinitySupport INSTANCE = null;
+    private static final AtomicBoolean INIT = new AtomicBoolean(false);
+
+    private final ILogger logger;
+    private ConcurrentMap<ThreadAffinity, ThreadAffinityController> registry = new ConcurrentHashMap<>();
+
+    private ThreadAffinitySupport(ILogger logger) {
+        this.logger = logger;
+    }
+
+    public ThreadAffinityController create(Class<? extends Thread> cls) {
+        checkNotNull(cls);
+        ThreadAffinity type = cls.getAnnotation(ThreadAffinity.class);
+        if (type != null) {
+            return getOrPutIfAbsent(registry, type, key -> new ThreadAffinityController(logger, type));
+        }
+
+        return null;
+    }
+
+    public static void init(ILogger logger) {
+        if (!isAffinityEnabled()) {
+            return;
+        }
+
+        if (!INIT.compareAndSet(false, true)) {
```

Comment:
I guess once set.... you will not be able to create a new HZ instance even if the old one is destroyed? I'm fine with that, but needs to be properly documented.

Reply:
I think we should be ok for the internal use, if we release this then we need to even re-think this strategy.